### PR TITLE
Support indicating partition transform when add column to Iceberg table

### PR DESCRIPTION
**Pull request by hantangwangd**
## Description

Through this PR, when we alter an Iceberg table to add a new column, we could also indicate it as a partition column. Moreover, we could use the `transform` functions and partition the table by the transformed value of the column, for example:

 - `ALTER TABLE iceberg.web.page_views ADD COLUMN zipcode VARCHAR WITH (partitioning = 'identity');`

 - `ALTER TABLE iceberg.web.page_views ADD COLUMN location VARCHAR WITH (partitioning = 'truncate(2)');`

 - `ALTER TABLE iceberg.web.page_views ADD COLUMN location VARCHAR WITH (partitioning = 'bucket(8)');`

Note: `Day`, `Month`, `Year`, `Hour` partition column transform functions are not supported in Presto Iceberg connector yet.

## Test Plan

 - Newly added IcebergDistributedTestBase.testAddPartitionColumn()

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== RELEASE NOTES ==

General Changes
* The Iceberg connector now supports `ALTER TABLE <table> ADD COLUMN <column> [WITH (partitioning = '<transform_func>')]` statements to indicate partition transform when add column to table
```

### CODE REVIEW

**Review by kiersten-stokes — APPROVED** [review]
This is working well for me! Nice work. I have two small comments about the docs below

File: presto-docs/src/main/sphinx/connector/iceberg.rst [hunk 1]
```diff
@@ -487,6 +487,20 @@ Alter table operations are supported in the connector::
 
      ALTER TABLE iceberg.web.page_views DROP COLUMN location;
 
+When adding a column, we can indicate it as a partition column, we can as well use the ``transform`` functions and partition
+the table by the transformed value of the column::
```
[the remark this answers is not in the document]
Comment (hantangwangd) [reply presto-docs/src/main/sphinx/connector/iceberg.rst]: Thanks, fixed.

File: presto-docs/src/main/sphinx/connector/iceberg.rst [hunk 2]
```diff
@@ -487,6 +487,20 @@ Alter table operations are supported in the connector::
...
+     ALTER TABLE iceberg.web.page_views ADD COLUMN zipcode VARCHAR WITH (partitioning = 'identity');
+
+     ALTER TABLE iceberg.web.page_views ADD COLUMN location VARCHAR WITH (partitioning = 'truncate(2)');
+
+     ALTER TABLE iceberg.web.page_views ADD COLUMN location VARCHAR WITH (partitioning = 'bucket(8)');
+
+.. note::
+
+    ``Day``, ``Month``, ``Year``, ``Hour`` partition column transform functions are not supported in Presto Iceberg
+        connector yet.
```
[the remark this answers is not in the document]
Comment (hantangwangd) [reply presto-docs/src/main/sphinx/connector/iceberg.rst]: Done.

File: presto-docs/src/main/sphinx/connector/iceberg.rst [hunk 3]
```diff
@@ -487,6 +487,19 @@ Alter table operations are supported in the connector::
 
      ALTER TABLE iceberg.web.page_views DROP COLUMN location;
 
+When adding a column, we can indicate it as a partition column. To do that, we should indicate the ``transform`` functions for it,
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/connector/iceberg.rst:490]:
```suggestion
To add a new column as a partition column, identify the transform functions for the column.
```

"We" should be avoided in technical documentation. For some explanations of why, see the documentation recommended word list entries for 'we' in [GitLab](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#we), [Google](https://developers.google.com/style/word-list#we), and [Microsoft](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/w/we).

File: presto-docs/src/main/sphinx/connector/iceberg.rst [hunk 4]
```diff
@@ -487,6 +487,19 @@ Alter table operations are supported in the connector::
...
+and then the table would be partitioned by the transformed value of the column::
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/connector/iceberg.rst:491]:
```suggestion
The table is partitioned by the transformed value of the column::
```

File: presto-docs/src/main/sphinx/connector/iceberg.rst [hunk 5]
```diff
@@ -487,6 +487,19 @@ Alter table operations are supported in the connector::
...
+
+     ALTER TABLE iceberg.web.page_views ADD COLUMN zipcode VARCHAR WITH (partitioning = 'identity');
+
+     ALTER TABLE iceberg.web.page_views ADD COLUMN location VARCHAR WITH (partitioning = 'truncate(2)');
+
+     ALTER TABLE iceberg.web.page_views ADD COLUMN location VARCHAR WITH (partitioning = 'bucket(8)');
+
+.. note::
+
+    ``Day``, ``Month``, ``Year``, ``Hour`` partition column transform functions are not supported in Presto Iceberg connector yet.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/connector/iceberg.rst:501]:
```suggestion
    ``Day``, ``Month``, ``Year``, ``Hour`` partition column transform functions are not supported in the Presto Iceberg connector.
```

Documentation should describe the current state of the product. "Yet" is an implied promise that this will, or is likely to happen that is out of scope for docs. Does that make sense?

**Review by steveburnett — CHANGES_REQUESTED** [review]
This looks good! I appreciate your attention to accurate formatting of the code. 

I made a few style and revision suggestions, and I explained my thoughts in the comments. If my suggested edits change the meaning of your text in a way you disagree with, let me know if I overlooked your intended meaning.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java [hunk 6]
```diff
@@ -144,6 +144,61 @@ public void testRenamePartitionColumn()
...
+
+        // Create partitioned table and insert some data
+        assertQuerySucceeds("create table add_partition_column(a int, b timestamp, c varchar) with (partitioning = ARRAY['b', 'truncate(c, 2)'])");
+        assertQuerySucceeds("insert into add_partition_column values(1, timestamp '1984-12-08 00:10:00', 'hw1001'), (2, timestamp '2001-01-08 12:10:01', 'hw1002')");
+        assertEquals(getQueryRunner().execute("SELECT count(*) FROM \"add_partition_column$partitions\"").getOnlyValue(), 2L);
+        assertEquals(getQueryRunner().execute("select row_count from \"add_partition_column$partitions\" where b = timestamp '1984-12-08 00:10:00' and c_trunc = 'hw'").getOnlyValue(), 1L);
+        assertEquals(getQueryRunner().execute("select row_count from \"add_partition_column$partitions\" where b = timestamp '2001-01-08 12:10:01' and c_trunc = 'hw'").getOnlyValue(), 1L);
+
+        // Add bucket partition column
+        assertQuerySucceeds("alter table add_partition_column add column d bigint with(partitioning = 'bucket(2)')");
```
Comment (ZacBlanco) [inline presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java:181]: Should we test for bad input values, e.g. `-1`? Does it error properly?

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java [hunk 7]
```diff
@@ -144,6 +144,61 @@ public void testRenamePartitionColumn()
...
+        assertQuerySucceeds("insert into add_partition_column values(1, timestamp '1984-12-08 00:10:00', 'hw1001', 1234), (1, timestamp '1984-12-08 00:10:00', 'hw1001', 1344)");
+        assertEquals(getQueryRunner().execute("SELECT count(*) FROM \"add_partition_column$partitions\"").getOnlyValue(), 4L);
+        assertEquals(getQueryRunner().execute("select row_count from \"add_partition_column$partitions\" where b = timestamp '1984-12-08 00:10:00' and c_trunc = 'hw' and d_bucket is null").getOnlyValue(), 1L);
```
Comment (ZacBlanco) [inline presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java:184]: So from my understanding the name of the columns in the `$partitions` table is based off of the name of the column with a suffix applied. In the case of bucketing, what if you had a column name with identity partitioning called `d_bucket` and a column `d` with partitioning function of `bucket(2)`. How would this case resolve when querying the table?

**Review by ZacBlanco — COMMENTED** [review]
Two questions, otherwise PR looks good!

File: presto-docs/src/main/sphinx/connector/iceberg.rst (hunk 5 above)
Comment (hantangwangd) [reply presto-docs/src/main/sphinx/connector/iceberg.rst:501]: I think we would support these transform functions sooner or later. So how could we express such a meaning? I see in the previous paragraph `Partition Column Transform`, it also use 'yet' to imply this promise, should we keep consistent with it?

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java (hunk 6 above)
Comment (hantangwangd) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java:181]: OK, I have added some test cases for this.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java (hunk 7 above)
Comment (hantangwangd) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/IcebergDistributedTestBase.java:184]: In situation of name clash, it would cause a fail when creating a table or add a column. Same partition column names are not allowed in one partition spec. I have added some test cases for this.

File: presto-docs/src/main/sphinx/connector/iceberg.rst (hunk 5 above)
Comment (steveburnett) [reply presto-docs/src/main/sphinx/connector/iceberg.rst:501]: > I think we would support these transform functions sooner or later. So how could we express such a meaning? I see in the previous paragraph `Partition Column Transform`, it also use 'yet' to imply this promise, should we keep consistent with it?

The simple answer to your first question is in docs, we don't. Technical documentation is not the place to express such a meaning. For references, see the Google developer documentation style guide about [present tense](https://developers.google.com/style/tense), and [Document future features](https://developers.google.com/style/future). 

About your second question, you make a good point about the other use of "yet" on this page, in `Partition Column Transform`! Thank you for noticing it. I was focused on the new text you were adding. 

Yes, to be consistent, that other use of "yet" in `Partition Column Transform` should also be removed. 

I hope this helps, let me know!

Comment (hantangwangd) [reply presto-docs/src/main/sphinx/connector/iceberg.rst]: Thanks for your patient explanation. I have fixed it.

**Review by tdcmeehan — APPROVED** [review]

File: presto-docs/src/main/sphinx/connector/iceberg.rst (hunk 5 above)
Comment (steveburnett) [reply presto-docs/src/main/sphinx/connector/iceberg.rst]: > Thanks for your patient explanation. I have fixed it, please take a look.

You're very welcome! It looks great.

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

Local build of docs, everything looks good.
